funnel/models/email_address.py

```diff
@@ -0,0 +1,572 @@
+        db.TIMESTAMP(timezone=True), nullable=False, default=db.func.utcnow()
+    )
+
+    #: Is this email address blocked from being used? If so, :attr:`email` should be
+    #: null. Blocks apply to the canonical address (without the +sub-address variation),
+    #: so a test for whether an address is blocked should use blake2b160_canonical to
+    #: load the record. Other records with the same canonical hash _may_ exist without
+    #: setting the flag due to a lack of database-side enforcement.
+    _is_blocked = db.Column('is_blocked', db.Boolean, nullable=False, default=False)
+
+    __table_args__ = (
+        db.CheckConstraint(
+            db.or_(
+                _is_blocked.isnot(True),
+                db.and_(_is_blocked.is_(True), email.is_(None),),
+            ),
+            'email_address_email_is_blocked_check',
+        ),
+    )
+
+    @hybrid_property
+    def is_blocked(self):
+        """"
+        Read-only flag indicating this email address is blocked from use. To set this
+        flag, call :classmethod:`mark_blocked` using the email address. The flag will be
+        simultaneously set on all matching instances.
+        """
+        return self._is_blocked
+
+    # This should not use `cached_property` as email is partially mutable
+    @property
+    def email_lower(self) -> Optional[str]:
+        """
+        Lowercase representation of the email address.
+        """
+        return self.email.lower() if self.email else None
+
+    # This should not use `cached_property` as email is partially mutable
+    @property
+    def email_canonical(self) -> Optional[str]:
+        """
+        Email address with the ``+sub-address`` portion of the mailbox removed.
+
+        This is only used to identify and prevent re-use of blocked email addresses
+        using the ``+sub-address`` method. Regular use does allow the ``+`` symbol.
+        Special handling for the gmail.com domain also strips periods from the
+        canonical representation. This makes the representation invalid for emailing.
+
+        The canonical representation is not stored, but its blake2b160 representation is
+        """
+        return canonical_email_representation(self.email)[0] if self.email else None
+
+    @with_roles(read={'all'})
+    @cached_property
+    def email_hash(self) -> str:
+        """Public identifier string for email address, usable in URLs."""
+        return base58.b58encode(self.blake2b160).decode()
+
+    @with_roles(call={'all'})
+    def md5(self) -> Optional[str]:
+        """MD5 hash of :property:`email_lower`, for legacy use only."""
+        return (
+            hashlib.md5(  # NOQA: S303 # skipcq: PTC-W1003 # nosec
+                self.email_lower.encode('utf-8')
+            ).hexdigest()
+            if self.email_lower
+            else None
+        )
+
+    def __str__(self) -> str:
+        """Cast email address into a string."""
+        return str(self.email or '')
```

Comment:
`self.email` is already a string, right? Why the `str()`?

Reply:
Fixed.